runtime/sema/check_composite_declaration.go

```diff
@@ -25,11 +25,74 @@ import (
 	"github.com/onflow/cadence/runtime/errors"
 )
 
+// while strictly speaking attachment declarations are not strictly the same thing as composite declarations,
+// they are subject to many of the same checks. In order to avoid duplicating these checks, we create a mapping
+// between an attachment declaration and its composite "equivalent" that we can use for these composite checks
+func (checker *Checker) attachmentAsComposite(declaration *ast.AttachmentDeclaration) *ast.CompositeDeclaration {
+	compositeDelcaration, ok := checker.Elaboration.AttachmentCompositeDeclarations[declaration]
```

Comment:
There is also potential to refactor this so that the composite checking/interpreting functions operate on a supertype of both attachments and composites (perhaps an `CompositeLikeDeclaration`) type. Once these are merged I will look into a follow-up PR to refactor this, potentially improving performance.

Reply:
https://github.com/onflow/cadence/pull/2148